### PR TITLE
Allow suid, exec, dev mount options to cancel nosuid/noexec/nodev

### PR DESCRIPTION
**Pull request by mheon**
Previously, we explicitly set noexec/nosuid/nodev on every mount, with no ability to disable them. The 'mount' command on Linux will accept their inverses without complaint, though - 'noexec' is counteracted by 'exec', 'nosuid' by 'suid', etc. Add support for passing these options at the command line to disable our explicit forcing of security options.

This also cleans up mount option handling significantly. We are still parsing options in more than one place, which isn't good, but option parsing for bind and tmpfs mounts has been unified.

Fixes: #3819 
Fixes: #3803

### CODE REVIEW

File: pkg/spec/spec.go [hunk 1]
```diff
@@ -394,43 +398,6 @@ func (config *CreateConfig) createConfigToOCISpec(runtime *libpod.Runtime, userM
 		}
 	}
 
-	// Make sure that the bind mounts keep options like nosuid, noexec, nodev.
-	mounts, err := pmount.GetMounts()
```
Comment (mheon) [inline pkg/spec/spec.go]: @rhatdan I think you originally wrote this, so you might want to take a look. I made adding `nosuid`/`noexec`/`nodev` unconditional (unless the user gives a `suid`/`exec`/`dev` to counter), not based on this pmounts call here. I think this is fine?

**Comment by mheon** [comment]
I'll add a few tests after lunch.

File: go.mod [hunk 2]
```diff
@@ -13,7 +13,7 @@ require (
 	github.com/containerd/continuity v0.0.0-20190426062206-aaeac12a7ffc // indirect
 	github.com/containernetworking/cni v0.7.1
 	github.com/containernetworking/plugins v0.8.1
-	github.com/containers/buildah v1.10.1
+	github.com/containers/buildah v1.8.4-0.20190821140209-376e52ee0142
```
Comment (mheon) [inline go.mod]: @vrothberg So I grabbed the top of buildah/master, and it appears to have downgraded, version-wise. We might want to look into that?

Comment (vrothberg) [reply go.mod]: The important part is behind the last dash (i.e., the commit `376e52ee0142`). It's a bug in go (no joke) which is trying to match a commit to a tag.

Commit 376e52ee0142 looks good.

File: pkg/spec/spec.go (hunk 1 above)
Comment (rhatdan) [reply pkg/spec/spec.go]: nosuid and nodev maybe.  Noexec could be an issue.  Does this affect volume mounts?  IE
If I volume mount in /usr/bin.  I probably don't want it marked as noexec.

File: pkg/spec/storage.go [hunk 3]
```diff
@@ -160,30 +160,23 @@ func (config *CreateConfig) parseVolumes(runtime *libpod.Runtime) ([]spec.Mount,
...
+		readonlyTmpfs := []string{"/tmp", "/var/tmp", "/run"}
+		options := []string{"rw", "rprivate", "exec", "nosuid", "nodev", "tmpcopyup"}
+		for _, dest := range readonlyTmpfs {
 			if _, ok := baseMounts[dest]; ok {
 				continue
 			}
 			localOpts := options
 			if dest == "/run" {
-				localOpts = append(localOpts, "noexec", "size=65536k")
+				localOpts = append(localOpts, "dev", "suid", "noexec", "size=65536k")
```
Comment (rhatdan) [inline pkg/spec/storage.go]: Why  are you doing this?

File: pkg/spec/storage.go [hunk 4]
```diff
@@ -775,16 +815,25 @@ func supercedeUserMounts(mounts []spec.Mount, configMount []spec.Mount) []spec.M
...
 		if m.Type == TypeBind {
-			m.Options = util.ProcessOptions(m.Options)
+			opts, err := util.ProcessOptions(m.Options, false)
+			if err != nil {
+				return nil, err
+			}
+			m.Options = opts
 		}
 		if m.Type == TypeTmpfs && filepath.Clean(m.Destination) != "/dev" {
-			m.Options = append(m.Options, "tmpcopyup")
```
Comment (rhatdan) [inline pkg/spec/storage.go]: Do you still have tmpcopyup by default for tmpfs?

**Comment by rhatdan** [comment]
You need to make sure that the mounts look the same with old podman as with this.  
Make sure 
podman run fedora mount

matches
./bin/podman run fedora mount

Then add a few other mounts using -v, --tmpfs and --mount and make sure everything is the same.

File: pkg/spec/storage.go (hunk 4 above)
Comment (mheon) [reply pkg/spec/storage.go]: Still defaulted

File: pkg/spec/storage.go (hunk 3 above)
Comment (mheon) [reply pkg/spec/storage.go]: Need to make sure that the parser doesn't see the mount is missing `nodev`/`nosuid` and add those

File: pkg/spec/spec.go (hunk 1 above)
Comment (mheon) [reply pkg/spec/spec.go]: Fair point. I'll poke at it more...

**Comment by mheon** [comment]
Test failures seem to be a mixed bag. It looks like my bind-mount options parsing code simplification over-simplified and broke things, but some of the rest seem like underlying bugs that are now being detected and reported.

**Comment by mheon** [comment]
Re-added functionality to infer mount options from those used on the base system.

**Comment by mheon** [comment]
I think tests will go green now. Still need to add a few more to cover new functionality.

**Comment by mheon** [comment]
Added tests. Good to go once this goes green.

**Comment by vrothberg** [comment]
I did a bigger chunks of review this morning. Maybe I'll find time this afternoon, latest tomorrow morning (EU).

File: test/e2e/run_volume_test.go [hunk 5]
```diff
@@ -162,4 +162,26 @@ var _ = Describe("Podman run with volumes", func() {
...
+
+		session = podmanTest.Podman([]string{"run", "--rm", "--tmpfs", "/run/test:suid,dev,exec", ALPINE, "grep", "/run/test", "/proc/self/mountinfo"})
+		session.WaitWithDefaultTimeout()
+		Expect(session.ExitCode()).To(Equal(0))
+		found, matches = session.GrepString("/run/test")
+		Expect(found).Should(BeTrue())
+		Expect(matches[0]).To(Not(ContainSubstring("noexec")))
+		Expect(matches[0]).To(Not(ContainSubstring("nodev")))
+		Expect(matches[0]).To(Not(ContainSubstring("nosuid")))
+	})
```
Comment (vrothberg) [inline test/e2e/run_volume_test.go]: Can we add an additional block box test?

A trivial one could be an executable `test.sh`:
```
#!/bin/sh
echo "Hello World!"
```

And then mount & execute it once with `exec` and grep for `Hello World!` and then mount & execute it with `noexec` and grep for ` Permission denied`.

File: go.mod (hunk 2 above)
Comment (vrothberg) [reply go.mod]: If we want to get rid of that, we can do `podman run --privileged -v `pwd`:/src:Z -it --rm golang:1.13-rc-buster` and `go get github.com/containers/buildah@376e52ee0142` inside the container. Go 1.13 will set the correct version/tag.

The annoying thing about that bug is that go mod's min-version algorithm can break here and downgrade a whole bunch of other dependencies.

File: pkg/spec/storage.go [hunk 6]
```diff
@@ -440,9 +431,23 @@ func getBindMount(args []string) (spec.Mount, error) {
 			} else {
 				return newMount, errors.Wrapf(optionArgError, "badly formatted option %q", val)
 			}
-		case "nosuid", "nodev", "noexec":
-			// TODO: detect duplication of these options.
-			// (Is this necessary?)
+		case "nosuid", "suid":
+			if setSuid {
+				return newMount, errors.Wrapf(optionArgError, "cannot pass 'nosuid' and 'suid' options more than once")
```
Comment (vrothberg) [inline pkg/spec/storage.go]: Clever trick to group them :+1:

File: pkg/spec/storage.go [hunk 7]
```diff
@@ -816,17 +817,46 @@ func supercedeUserMounts(mounts []spec.Mount, configMount []spec.Mount) []spec.M
...
 // Ensure mount options on all mounts are correct
 func initFSMounts(inputMounts []spec.Mount) ([]spec.Mount, error) {
+	// We need to look up mounts so we can figure out the proper mount flags
+	// to apply.
+	systemMounts, err := pmount.GetMounts()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error retrieving system mounts to look up mount options")
+	}
+
+	// TODO: We probably don't need to re-build the mounts array
```
Comment (vrothberg) [inline pkg/spec/storage.go]: Nit: I think so too. Operating on `inputMounts[i]` should work.

**Comment by TomSweeneyRedHat** [comment]
LGTM, nothing beyond @vrothbergs comments.

File: test/e2e/run_volume_test.go (hunk 5 above)
Comment (mheon) [reply test/e2e/run_volume_test.go]: Done - I just mounted in `/bin` from the host, but it should be a sufficient test.

File: pkg/spec/storage.go (hunk 7 above)
Comment (mheon) [reply pkg/spec/storage.go]: I'll leave it off for now, this is large enough already

**Comment by mheon** [comment]
Should be ready for merge.
@baude @rhatdan @vrothberg @giuseppe PTAL

**Pull request merged**